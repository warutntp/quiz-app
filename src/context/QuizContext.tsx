import React, {
  createContext,
  useReducer,
  ReactNode,
  Dispatch,
  useEffect,
  useMemo,
} from "react";
import { QuizModels } from "../types/QuizModels";
import { LeaderModels } from "../types/LeaderModels";
import { RandomData } from "../utils/RandomData";
import { QuizData } from "../data/QuizData";
import {
  loadLeaderDataFromLocalStorage,
  saveLeaderDataToLocalStorage,
} from "../utils/LocalStorage";

type State = {
  currentQuestion: QuizModels.QuizModel | null;
  leaderBoard: LeaderModels.LeaderModel[];
  name: string;
  score: number;
  isQuizActive: boolean;
  questionIndex: number;
  totalQuestions: QuizModels.QuizModel[];
  answers: string[];
};

const initialState: State = {
  currentQuestion: null,
  leaderBoard: [],
  name: "",
  score: 0,
  isQuizActive: false,
  questionIndex: 0,
  totalQuestions: [],
  answers: [],
};

type Action =
  | { type: "SET_NAME"; payload: string }
  | { type: "START_QUIZ" }
  | { type: "ANSWER_QUESTION"; payload: string }
  | { type: "NEXT_QUESTION" }
  | { type: "PREVIOUS_QUESTION" }
  | { type: "SUBMIT_QUIZ" }
  | { type: "LOAD_LEADERBOARD"; payload: LeaderModels.LeaderModel[] };

const quizReducer = (state: State, action: Action): State => {
  switch (action.type) {
    case "SET_NAME":
      return { ...state, name: action.payload };
    case "START_QUIZ": {
      const shuffled = RandomData([...QuizData], 20);
      return {
        ...state,
        isQuizActive: true,
        totalQuestions: shuffled,
        currentQuestion: shuffled[0],
        questionIndex: 0,
        answers: Array(shuffled.length).fill(null),
      };
    }
    case "ANSWER_QUESTION": {
      const newAnswers = [...state.answers];
      newAnswers[state.questionIndex] = action.payload;
      return { ...state, answers: newAnswers };
    }
    case "NEXT_QUESTION":
      if (state.questionIndex < state.totalQuestions.length - 1) {
        return {
          ...state,
          questionIndex: state.questionIndex + 1,
          currentQuestion: state.totalQuestions[state.questionIndex + 1],
        };
      }
      return state;
    case "PREVIOUS_QUESTION":
      if (state.questionIndex > 0) {
        return {
          ...state,
          questionIndex: state.questionIndex - 1,
          currentQuestion: state.totalQuestions[state.questionIndex - 1],
        };
      }
      return state;
    case "SUBMIT_QUIZ": {
      const newScore = state.answers.reduce((acc, answer, idx) => {
        if (answer === state.totalQuestions[idx].correct) {
          return acc + 1;
        }
        return acc;
      }, 0);

      const newLeaderBoard = [
        ...state.leaderBoard,
        { name: state.name, score: newScore },
      ];
      saveLeaderDataToLocalStorage(newLeaderBoard);

      return {
        ...state,
        score: newScore,
        isQuizActive: false,
        leaderBoard: newLeaderBoard,
      };
    }
    case "LOAD_LEADERBOARD":
      return { ...state, leaderBoard: action.payload };
    default:
      return state;
  }
};

interface QuizContextProps {
  state: State;
  dispatch: Dispatch<Action>;
}

const QuizContext = createContext<QuizContextProps>(undefined as any);

const QuizProvider: React.FC<{ children?: ReactNode }> = ({ children }) => {
  const [state, dispatch] = useReducer(quizReducer, initialState);

  useEffect(() => {
    const savedLeaderBoard = loadLeaderDataFromLocalStorage();
    dispatch({ type: "LOAD_LEADERBOARD", payload: savedLeaderBoard });
  }, [dispatch]);

  const contextValue = useMemo(() => ({ state, dispatch }), [state, dispatch]);

  return (
    <QuizContext.Provider value={contextValue}>{children}</QuizContext.Provider>
  );
};

export { QuizContext, QuizProvider };
