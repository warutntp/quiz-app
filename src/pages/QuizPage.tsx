import React, { useContext } from "react";
import { QuizContext } from "../context/QuizContext";
import QuizShow from "../components/QuizShow";
import LeaderBoard from "../components/LeaderBoard";
import QuizForm from "../components/QuizForm";
import QuizNavigation from "../components/QuizNavigation";

const QuizPage = () => {
  const context = useContext(QuizContext);

  if (!context) {
    throw new Error("QuizContext must be used within a QuizProvider");
  }

  const { state, dispatch } = context;
  const handleSubmitName = (e: React.FormEvent) => {
    e.preventDefault();
    dispatch({ type: "START_QUIZ" });
  };

  const handleAnswer = (answer: string) => {
    dispatch({ type: "ANSWER_QUESTION", payload: answer });
  };

  const handleNext = () => {
    dispatch({ type: "NEXT_QUESTION" });
  };

  const handleBack = () => {
    dispatch({ type: "PREVIOUS_QUESTION" });
  };

  const handleSubmitQuiz = () => {
    if (state.answers.some((answer) => answer === null)) {
      alert("Please answer all questions before submitting.");
      return;
    }
    dispatch({ type: "SUBMIT_QUIZ" });
  };

  return (
    <div className="p-4 max-w-lg mx-auto">
      <h1 className="text-center text-2xl font-bold mb-4">Quiz App</h1>
      {!state.isQuizActive && (
        <QuizForm
          name={state.name}
          setName={(name) => dispatch({ type: "SET_NAME", payload: name })}
          handleSubmitName={handleSubmitName}
        />
      )}
      {state.isQuizActive && state.currentQuestion && (
        <>
          <p className="mt-4">
            Question {state.questionIndex + 1} of{" "}
            {state.shuffledQuestions.length}
          </p>
          <QuizShow
            question={state.currentQuestion.question}
            answers={state.currentQuestion.answers}
            handleAnswer={handleAnswer}
            selectedAnswer={state.answers[state.questionIndex]}
          />
          <QuizNavigation
            questionIndex={state.questionIndex}
            shuffledQuestions={state.shuffledQuestions.length}
            answers={state.answers}
            handleNext={handleNext}
            handleBack={handleBack}
            handleSubmitQuiz={handleSubmitQuiz}
          />
        </>
      )}
      {!state.isQuizActive && state.score > 0 && (
        <div className="mt-4 p-4 border rounded-lg shadow-md bg-white">
          <h2 className="text-lg font-bold mb-4">Your Score: {state.score}</h2>
        </div>
      )}
      <LeaderBoard scores={state.leaderBoard} />
    </div>
  );
};

export default QuizPage;
