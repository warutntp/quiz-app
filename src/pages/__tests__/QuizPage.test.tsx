import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import QuizPage from "../QuizPage";
import { QuizProvider } from "../../context/QuizContext";
import "@testing-library/jest-dom/extend-expect";

test("QuizPage", () => {
  const { container } = render(
    <QuizProvider>
      <QuizPage />
    </QuizProvider>
  );
  expect(container).toMatchSnapshot();
});

test("renders QuizPage and starts quiz correctly", async () => {
  render(
    <QuizProvider>
      <QuizPage />
    </QuizProvider>
  );

  const input = screen.getByLabelText("Enter your name:");
  fireEvent.change(input, { target: { value: "John" } });

  const button = screen.getByText("Start Quiz");
  fireEvent.click(button);

  expect(
    screen.getByText((content, element) => content.startsWith("Question 1 of"))
  ).toBeInTheDocument();
});
