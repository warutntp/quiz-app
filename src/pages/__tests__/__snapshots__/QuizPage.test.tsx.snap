// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App 1`] = `
<div>
  <div
    class="p-4 max-w-lg mx-auto"
  >
    <h1
      class="text-center text-2xl font-bold mb-4"
    >
      Quiz App
    </h1>
    <form
      class="mb-4"
    >
      <div
        class="mb-2"
      >
        <label
          class="block text-sm font-medium text-gray-700"
          for="name"
        >
          Enter your name:
        </label>
        <input
          class="mt-1 p-2 block w-full border border-gray-300 rounded-md shadow-sm"
          id="name"
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="w-full py-2 px-4 bg-blue-500 text-white rounded hover:bg-blue-700"
        type="submit"
      >
        Start Quiz
      </button>
    </form>
    <div
      class="p-4 border rounded-lg shadow-md bg-white mt-4 max-h-[500px] overflow-y-auto"
    >
      <h2
        class="text-lg font-bold mb-4"
      >
        LeaderBoard
      </h2>
      <ul />
    </div>
  </div>
</div>
`;

exports[`QuizPage 1`] = `
<div>
  <div
    class="p-4 max-w-lg mx-auto"
  >
    <h1
      class="text-center text-2xl font-bold mb-4"
    >
      Quiz App
    </h1>
    <form
      class="mb-4"
    >
      <div
        class="mb-2"
      >
        <label
          class="block text-sm font-medium text-gray-700"
          for="name"
        >
          Enter your name:
        </label>
        <input
          class="mt-1 p-2 block w-full border border-gray-300 rounded-md shadow-sm"
          id="name"
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="w-full py-2 px-4 bg-blue-500 text-white rounded hover:bg-blue-700"
        type="submit"
      >
        Start Quiz
      </button>
    </form>
    <div
      class="p-4 border rounded-lg shadow-md bg-white mt-4 max-h-[500px] overflow-y-auto"
    >
      <h2
        class="text-lg font-bold mb-4"
      >
        LeaderBoard
      </h2>
      <ul />
    </div>
  </div>
</div>
`;
